### Global Superstone Sales Analysis

### *Tasks:* 
#### Data Preparation
Import the dataset into your tool of choice and clean any 
inconsistencies. 

Descriptive Analysis
Summarize total sales, average profit, and number of 
orders by region or category. 

#### Visualizations
▪ Create line charts showing sales trends over time. 

▪ Generate bar or column charts comparing total sales by category and 
region. 

▪ Use heatmaps or scatter plots to show correlations between profit and 
quantity sold.

▪ Develop a dashboard using dash and deploy it to clearly communicate all unveiled insights

### Insights 
▪ Identify the region with the highest sales and profit. 

▪ Highlight the most profitable and least profitable product categories. 

▪ Any other insights. 
 

* Installing ploty packages 

In [39]:
# installing plotly
#%pip install plotly

* Importing the required packages

In [40]:
# importing the required modules
import pandas as pd  # for data manipulation
import numpy as np  # for numerical computation
import matplotlib.pyplot as plt # for data visualization
import seaborn as sns # for data visualization 
import plotly.express as px # for data visualization
import plotly.graph_objects as go # for data visualization
from plotly.subplots import make_subplots # for data visualization
import os # for file handling
import sys # for system operations
import warnings # for ignoring warnings
warnings.filterwarnings('ignore') # ignoring warnings


* Laoding the dataset

In [41]:
# Load the data
df = pd.read_csv("C:\\Users\\Admin\\OneDrive\\R STA1040\\superstore.csv")
df # display the first five rows of the dataframe

Category         City        Country Customer.ID  \
0      Office Supplies  Los Angeles  United States   LS-172304   
1      Office Supplies  Los Angeles  United States   MV-174854   
2      Office Supplies  Los Angeles  United States   CS-121304   
3      Office Supplies  Los Angeles  United States   CS-121304   
4      Office Supplies  Los Angeles  United States   AP-109154   
...                ...          ...            ...         ...   
51285  Office Supplies  Los Angeles  United States   AM-103604   
51286  Office Supplies  Los Angeles  United States   AM-103604   
51287  Office Supplies  Los Angeles  United States   HR-147704   
51288  Office Supplies  Los Angeles  United States   RM-196754   
51289  Office Supplies  Los Angeles  United States   FH-143654   

          Customer.Name  Discount Market  记录数               Order.Date  \
0      Lycoris Saunders       0.0     US    1  2011-01-07 00:00:00.000   
1         Mark Van Huff       0.0     US    1  2011-01-21 00:00:00.000   
2          Chad Sievert       0.0     US    1  2011-08-05 00:00:00.000   
3          Chad Sievert       0.0     US    1  2011-08-05 00:00:00.000   
4        Arthur Prichep       0.0     US    1  2011-09-29 00:00:00.000   
...                 ...       ...    ...  ...                      ...   
51285    Alice McCarthy       0.2     US    1  2014-12-03 00:00:00.000   
51286    Alice McCarthy       0.2     US    1  2014-12-03 00:00:00.000   
51287    Hallie Redmond       0.2     US    1  2014-12-18 00:00:00.000   
51288     Robert Marley       0.2     US    1  2014-12-25 00:00:00.000   
51289      Fred Hopkins       0.2     US    1  2014-12-26 00:00:00.000   

             Order.ID  ... Sales      Segment                Ship.Date  \
0      CA-2011-130813  ...    19     Consumer  2011-01-09 00:00:00.000   
1      CA-2011-148614  ...    19     Consumer  2011-01-26 00:00:00.000   
2      CA-2011-118962  ...    21     Consumer  2011-08-09 00:00:00.000   
3      CA-2011-118962  ...   111     Consumer  2011-08-09 00:00:00.000   
4      CA-2011-146969  ...     6     Consumer  2011-10-03 00:00:00.000   
...               ...  ...   ...          ...                      ...   
51285  CA-2014-109701  ...    69    Corporate  2014-12-04 00:00:00.000   
51286  CA-2014-109701  ...     9    Corporate  2014-12-04 00:00:00.000   
51287  CA-2014-106964  ...    12  Home Office  2014-12-21 00:00:00.000   
51288  CA-2014-145219  ...    90  Home Office  2014-12-26 00:00:00.000   
51289  CA-2014-121398  ...   154    Corporate  2014-12-30 00:00:00.000   

            Ship.Mode  Shipping.Cost       State  Sub.Category  Year  \
0        Second Class           4.37  California         Paper  2011   
1      Standard Class           0.94  California         Paper  2011   
2      Standard Class           1.81  California         Paper  2011   
3      Standard Class           4.59  California         Paper  2011   
4      Standard Class           1.32  California         Paper  2011   
...               ...            ...         ...           ...   ...   
51285        Same Day           5.15  California       Binders  2014   
51286        Same Day           0.44  California       Binders  2014   
51287     First Class           0.31  California       Binders  2014   
51288     First Class          15.95  California       Binders  2014   
51289  Standard Class           9.59  California       Binders  2014   

             Market2 weeknum  
0      North America       2  
1      North America       4  
2      North America      32  
3      North America      32  
4      North America      40  
...              ...     ...  
51285  North America      49  
51286  North America      49  
51287  North America      51  
51288  North America      52  
51289  North America      52  

[51290 rows x 27 columns]

In [42]:
# converting the column name 记录数 to NO of Records
df.rename(columns={'记录数':'No.Records'}, inplace=True)

# Converting Ship.Date and Order.Date to datetime format
df['Ship.Date'] = pd.to_datetime(df['Ship.Date'])
df['Order.Date'] = pd.to_datetime(df['Order.Date'])

df.head()

Category         City        Country Customer.ID     Customer.Name  \
0  Office Supplies  Los Angeles  United States   LS-172304  Lycoris Saunders   
1  Office Supplies  Los Angeles  United States   MV-174854     Mark Van Huff   
2  Office Supplies  Los Angeles  United States   CS-121304      Chad Sievert   
3  Office Supplies  Los Angeles  United States   CS-121304      Chad Sievert   
4  Office Supplies  Los Angeles  United States   AP-109154    Arthur Prichep   

   Discount Market  No.Records Order.Date        Order.ID  ... Sales  \
0       0.0     US           1 2011-01-07  CA-2011-130813  ...    19   
1       0.0     US           1 2011-01-21  CA-2011-148614  ...    19   
2       0.0     US           1 2011-08-05  CA-2011-118962  ...    21   
3       0.0     US           1 2011-08-05  CA-2011-118962  ...   111   
4       0.0     US           1 2011-09-29  CA-2011-146969  ...     6   

    Segment  Ship.Date       Ship.Mode  Shipping.Cost       State  \
0  Consumer 2011-01-09    Second Class           4.37  California   
1  Consumer 2011-01-26  Standard Class           0.94  California   
2  Consumer 2011-08-09  Standard Class           1.81  California   
3  Consumer 2011-08-09  Standard Class           4.59  California   
4  Consumer 2011-10-03  Standard Class           1.32  California   

   Sub.Category  Year        Market2 weeknum  
0         Paper  2011  North America       2  
1         Paper  2011  North America       4  
2         Paper  2011  North America      32  
3         Paper  2011  North America      32  
4         Paper  2011  North America      40  

[5 rows x 27 columns]

### The dataset contains 27 columns and 51290 rows:
category: The category of products sold in the superstore.

city: The city where the order was placed.

country: The country in which the superstore is located.

customer_id: A unique identifier for each customer.

customer_name: The name of the customer who placed the order.

discount: The discount applied to the order.

market: The market or region where the superstore operates.

ji_lu_shu: An unknown or unspecified column.

order_date: The date when the order was placed.

order_id: A unique identifier for each order.

order_priority: The priority level of the order.

product_id: A unique identifier for each product.

product_name: The name of the product.

profit: The profit generated from the order.

quantity: The quantity of products ordered.

region: The region where the order was placed.

row_id: A unique identifier for each row in the dataset.

sales: The total sales amount for the order.

segment: The customer segment (e.g., consumer, corporate, or home office).

ship_date: The date when the order was shipped.

ship_mode: The shipping mode used for the order.

shipping_cost: The cost of shipping for the order.

state: The state or region within the country.

sub_category: The sub-category of products within the main category.

year: The year in which the order was placed.

market2: Another column related to market information.

weeknum: The week number when the order was placed.

This dataset can be used for various data analysis tasks, including understanding sales patterns, customer behavior, and profitability in the context of a global superstore.


* Data undertanding and cleaning

In [43]:
# Data understanding
df.info() # display the information about the dataframe


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 27 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Category        51290 non-null  object        
 1   City            51290 non-null  object        
 2   Country         51290 non-null  object        
 3   Customer.ID     51290 non-null  object        
 4   Customer.Name   51290 non-null  object        
 5   Discount        51290 non-null  float64       
 6   Market          51290 non-null  object        
 7   No.Records      51290 non-null  int64         
 8   Order.Date      51290 non-null  datetime64[ns]
 9   Order.ID        51290 non-null  object        
 10  Order.Priority  51290 non-null  object        
 11  Product.ID      51290 non-null  object        
 12  Product.Name    51290 non-null  object        
 13  Profit          51290 non-null  float64       
 14  Quantity        51290 non-null  int64         
 15  Re

In [44]:
df.describe() # display the summary statistics of the dataframe


Discount  No.Records        Profit      Quantity       Row.ID  \
count  51290.000000     51290.0  51290.000000  51290.000000  51290.00000   
mean       0.142908         1.0     28.610982      3.476545  25645.50000   
std        0.212280         0.0    174.340972      2.278766  14806.29199   
min        0.000000         1.0  -6599.978000      1.000000      1.00000   
25%        0.000000         1.0      0.000000      2.000000  12823.25000   
50%        0.000000         1.0      9.240000      3.000000  25645.50000   
75%        0.200000         1.0     36.810000      5.000000  38467.75000   
max        0.850000         1.0   8399.976000     14.000000  51290.00000   

              Sales  Shipping.Cost          Year       weeknum  
count  51290.000000   51290.000000  51290.000000  51290.000000  
mean     246.498440      26.375818   2012.777208     31.287112  
std      487.567175      57.296810      1.098931     14.429795  
min        0.000000       0.002000   2011.000000      1.000000  
25%       31.000000       2.610000   2012.000000     20.000000  
50%       85.000000       7.790000   2013.000000     33.000000  
75%      251.000000      24.450000   2014.000000     44.000000  
max    22638.000000     933.570000   2014.000000     53.000000

In [45]:
#df.shape # display the shape of the dataframe
df.dtypes # display the data types of the dataframe


Category                  object
City                      object
Country                   object
Customer.ID               object
Customer.Name             object
Discount                 float64
Market                    object
No.Records                 int64
Order.Date        datetime64[ns]
Order.ID                  object
Order.Priority            object
Product.ID                object
Product.Name              object
Profit                   float64
Quantity                   int64
Region                    object
Row.ID                     int64
Sales                      int64
Segment                   object
Ship.Date         datetime64[ns]
Ship.Mode                 object
Shipping.Cost            float64
State                     object
Sub.Category              object
Year                       int64
Market2                   object
weeknum                    int64
dtype: object

In [46]:

df.isnull().sum() # display the missing values in the dataframe

Category          0
City              0
Country           0
Customer.ID       0
Customer.Name     0
Discount          0
Market            0
No.Records        0
Order.Date        0
Order.ID          0
Order.Priority    0
Product.ID        0
Product.Name      0
Profit            0
Quantity          0
Region            0
Row.ID            0
Sales             0
Segment           0
Ship.Date         0
Ship.Mode         0
Shipping.Cost     0
State             0
Sub.Category      0
Year              0
Market2           0
weeknum           0
dtype: int64

In [47]:
df.nunique() # display the unique values in the dataframe

Category              3
City               3636
Country             147
Customer.ID        4873
Customer.Name       795
Discount             27
Market                7
No.Records            1
Order.Date         1430
Order.ID          25035
Order.Priority        4
Product.ID        10292
Product.Name       3788
Profit            24575
Quantity             14
Region               13
Row.ID            51290
Sales              2246
Segment               3
Ship.Date          1464
Ship.Mode             4
Shipping.Cost     16877
State              1094
Sub.Category         17
Year                  4
Market2               6
weeknum              53
dtype: int64

In [48]:
# Check for duplicates
df.duplicated().sum() # display the duplicates in the dataframe

0

#### Descriptive Analysis

#### Summarize by Region

#### Summarize by Category

In [49]:
#Summarize total sales, average profit, and number of orders by region or category.
region_summary = df.groupby('Region').agg({'Sales': 'sum', 'Profit': 'mean', 'Quantity': 'sum'}).reset_index()
print("Region Summary:")
region_summary.sort_values(by='Sales', ascending=False)


Region Summary:


Region    Sales     Profit  Quantity
3          Central  2822399  28.011512     41762
10           South  1600960  21.122011     25206
7            North  1248192  40.668329     18021
9          Oceania  1100207  34.439091     12838
11  Southeast Asia   884438   5.705442     11822
8       North Asia   848349  70.820539      8851
5             EMEA   806184   8.728966     11517
0           Africa   783776  19.374674     10564
4     Central Asia   752839  64.687591      7715
12            West   725514  33.849032     12266
6             East   678834  32.135808     10618
2        Caribbean   324281  20.456403      6299
1           Canada    66932  46.399453       833

In [50]:
# Visualizing the Region Summary
fig = make_subplots(rows=1, cols=3, subplot_titles=("Total Sales by Region", "Average Profit by Region", "Total Quantity by Region"))

fig.add_trace(go.Bar(x=region_summary['Region'], y=region_summary['Sales'], marker=dict(color=region_summary['Sales'], coloraxis="coloraxis")), row=1, col=1)
fig.add_trace(go.Bar(x=region_summary['Region'], y=region_summary['Profit'], marker=dict(color=region_summary['Profit'], coloraxis="coloraxis")), row=1, col=2)
fig.add_trace(go.Bar(x=region_summary['Region'], y=region_summary['Quantity'], marker=dict(color=region_summary['Quantity'], coloraxis="coloraxis")), row=1, col=3)

fig.update_layout(coloraxis=dict(colorscale='Viridis'), showlegend=False, title_text="Region Summary")

# update the background color of the plot not the bars
fig.update_layout(plot_bgcolor='#50C878', paper_bgcolor='#f5f5f5', font_color="black")
fig.show()



In [51]:
category_summary = df.groupby('Category').agg({'Sales': 'sum', 'Profit': 'mean', 'Quantity': 'count'}).reset_index()
print("Category Summary:")
category_summary.sort_values(by='Sales', ascending=False)


Category Summary:


Category    Sales     Profit  Quantity
2       Technology  4744691  65.454958     10141
0        Furniture  4110884  28.878567      9876
1  Office Supplies  3787330  16.578961     31273

In [52]:
# Using a donut chart to visualize the category summary by sales and profit
# indicate which chart is for sales and which chart is for profit by using different colors and labels
fig1 = make_subplots(rows=1, cols=2, specs=[[{'type':'domain'}, {'type':'domain'}]])
fig1.add_trace(go.Pie(labels=category_summary['Category'], values=category_summary['Sales'], name='Sales', marker_colors=px.colors.sequential.RdPu_r), 1, 1)
fig1.add_trace(go.Pie(labels=category_summary['Category'], values=category_summary['Profit'], name='Profit', marker_colors=px.colors.sequential.RdPu_r), 1, 2)
fig1.update_traces(hole=.5, hoverinfo="label+percent+name")
fig1.update_layout(title_text="Category Summary by Sales and Profit")
fig1.update_layout(paper_bgcolor="lightblue")
fig1.show()


In [53]:
# Creating a subplot for plotting the sales, profit and quantity by region on pie chart
fig2 = make_subplots(rows=1, cols=3, specs=[[{'type':'domain'}, {'type':'domain'}, {'type':'domain'}]], subplot_titles=('Sales by Region', 'Profit by Region', 'Quantity by Region'))

# Plotting the sales by region

fig2.add_trace(go.Pie(labels=region_summary['Region'], values=region_summary['Sales'], name='Sales'), 1, 1)

# Plotting the profit by region
fig2.add_trace(go.Pie(labels=region_summary['Region'], values=region_summary['Profit'], name='Profit'), 1, 2)

# Plotting the quantity by region
fig2.add_trace(go.Pie(labels=region_summary['Region'], values=region_summary['Quantity'], name='Quantity'), 1, 3)

fig2.update_layout(title_text='Sales, Profit and Quantity by Region')
fig2.show()


### Sales Trend over time

In [54]:
import plotly.express as px

#df['order_date'] = pd.to_datetime(df['order_date'])
df['Year_month'] = df['Order.Date'].dt.to_period('M').astype(str)
sales_trends = df.groupby('Year_month')['Sales'].sum().reset_index()

fig3 = px.line(sales_trends, x='Year_month', y='Sales', title='Sales Trends Over Time',
               labels={'Year_month': 'Year-Month', 'Sales': 'Total Sales'},
               markers=True)
fig3.update_traces(line_color='blue')
fig3.show()


In [55]:
# Create Year_month column
df['Year_month'] = df['Order.Date'].dt.to_period('M').astype(str)

# Region wise sales trends
region_sales_trends = df.groupby(['Year_month', 'Region'])['Sales'].sum().reset_index()

fig4 = px.line(region_sales_trends, x='Year_month', y='Sales', title='Region Wise Sales Trends Over Time',
                color='Region', labels={'Year_month': 'Year-Month', 'Sales': 'Total Sales'},
                markers=True)
fig4.show()

In [56]:
# Region wise sales, profit and quantity trends over time in subplots for top 5 regions
top_5_regions = df.groupby('Region')['Sales'].sum().sort_values(ascending=False).head(3).index
region_summary = df.groupby(['Year_month', 'Region']).agg({'Sales': 'sum', 'Profit': 'sum', 'Quantity': 'sum'}).reset_index()

fig5 = make_subplots(rows=1, cols=2, subplot_titles=('Sales Trends', 'Profit Trends'))

# Plotting the sales trends
for region in top_5_regions:
    fig5.add_trace(go.Scatter
                        (x=region_summary[region_summary['Region'] == region]['Year_month'],
                        y=region_summary[region_summary['Region'] == region]['Sales'],
                        mode='lines+markers', name=region), row=1, col=1)
    
# Plotting the profit trends
for region in top_5_regions:
    fig5.add_trace(go.Scatter
                        (x=region_summary[region_summary['Region'] == region]['Year_month'],
                        y=region_summary[region_summary['Region'] == region]['Profit'],
                        mode='lines+markers', name=region), row=1, col=2)
    
# Plotting the quantity trends
#for region in top_5_regions:
    #fig4.add_trace(go.Scatter
                        #(x=region_summary[region_summary['Region'] == region]['Year_month'],
                        #y=region_summary[region_summary['Region'] == region]['Quantity'],
                        #mode='lines+markers', name=region), row=2, col=1)

fig5.update_layout(title_text='Top 3 Region Wise Sales, Profit and Quantity Trends Over Time')
fig5.show()

### Total sales by Category and region

In [57]:
# Using bar chart to visualize the top 10 products by sales and profit and quantity using a subplot
top_5_products = df.groupby('Product.Name').agg({'Sales': 'sum', 'Profit': 'sum', 'Quantity': 'sum'}).reset_index()
top_5_products = top_5_products.sort_values(by='Sales', ascending=False).head(3)

fig6 = make_subplots(rows=1, cols=3, subplot_titles=('Sales by Product', 'Profit by Product', 'Quantity by Product'))

# Plotting the sales by product
fig6.add_trace(go.Bar(x=top_5_products['Product.Name'], y=top_5_products['Sales'], name='Sales'), 1, 1)

# Plotting the profit by product
fig6.add_trace(go.Bar(x=top_5_products['Product.Name'], y=top_5_products['Profit'], name='Profit'), 1, 2)

# Plotting the quantity by product
fig6.add_trace(go.Bar(x=top_5_products['Product.Name'], y=top_5_products['Quantity'], name='Quantity'), 1, 3)

fig6.update_layout(title_text='Top 5 Products by Sales, Profit and Quantity')
# background color
fig6.update_layout(paper_bgcolor="khaki")
fig6.show()


In [58]:
# Total sales, profit and quantity by segment
segment_summary = df.groupby('Segment').agg({'Sales': 'sum', 'Profit': 'sum', 'Quantity': 'sum'}).reset_index()
segment_summary

Segment    Sales        Profit  Quantity
0     Consumer  6508141  749239.78206     92157
1    Corporate  3824808  441208.32866     53565
2  Home Office  2309956  277009.18056     32590

In [59]:
# A plot of total sales and profit by segment as grouped bar chart on a single plot
fig7 = go.Figure(data=[
    go.Bar(name='Sales', x=segment_summary['Segment'], y=segment_summary['Sales']),
    go.Bar(name='Profit', x=segment_summary['Segment'], y=segment_summary['Profit'])
])

# Change the bar mode
fig7.update_layout(barmode='group')
fig7.update_layout(title_text='Total Sales and Profit by Segment')
fig7.show()

In [60]:
# Total sales, profit and quantity by Region, Segment and Category after grouping by Region, Segment and Category
region_segment_category_summary = df.groupby(['Region']).agg({'Sales': 'sum', 'Profit': 'sum'}).reset_index()
region_segment_category_summary


Region    Sales        Profit
0           Africa   783776   88871.63100
1           Canada    66932   17817.39000
2        Caribbean   324281   34571.32104
3          Central  2822399  311403.98164
4     Central Asia   752839  132480.18700
5             EMEA   806184   43897.97100
6             East   678834   91522.78000
7            North  1248192  194597.95252
8       North Asia   848349  165578.42100
9          Oceania  1100207  120089.11200
10           South  1600960  140355.76618
11  Southeast Asia   884438   17852.32900
12            West   725514  108418.44890

In [61]:
# A plot of total sales and profit by segment, Region, and Category as grouped bar chart on a single plot
fig8 = go.Figure(data=[
    go.Bar(name='Sales', x=region_segment_category_summary['Region'], y=region_segment_category_summary['Sales']),
    go.Bar(name='Profit', x=region_segment_category_summary['Region'], y=region_segment_category_summary['Profit'])
])

# Change the bar mode
fig8.update_layout(barmode='group')
fig8.update_layout(title_text='Total Sales and Profit by Region')
fig8.show()

In [62]:
# Sales by Region and Category using a bar chart and a grouped bar chart and arrange them in a subplot
region_category_summary = df.groupby(['Region', 'Category']).agg({'Sales': 'sum'}).reset_index()

fig9 = make_subplots(rows=1, cols=2, subplot_titles=('Sales by Region and Category', 'Sales by Region and Category'))

# Plotting the sales by region and category
fig9.add_trace(go.Bar(x=region_category_summary['Region'], y=region_category_summary['Sales'], name='Sales'), 1, 1)

# Plotting the sales by region and category as grouped bar chart
fig9.add_trace(go.Bar(x=region_category_summary['Region'], y=region_category_summary['Sales'], name='Sales'), 1, 2)

fig9.update_layout(title_text='Sales by Region and Category')
fig9.show()


### Correlation between Profit and Quantity Sold

## HeatMap

In [88]:
# Using heatmap to visualize the correlation between sales, profit, and quantity, discount, and shipping cost
correlation = df[['Sales', 'Profit','Discount']].corr()
fig10 = go.Figure(data=go.Heatmap(z=correlation.values, x=correlation.columns, y=correlation.columns))
fig10.update_layout(title_text='Correlation between Sales, Profit, Discount')
fig10.show()

In [64]:
# Plotting a heatmap to visualize the sales, profit and quantity by Region and Category
region_category_summary = df.groupby(['Region', 'Category']).agg({'Sales': 'sum', 'Profit': 'sum', 'Quantity': 'sum'}).reset_index()

fig11 = go.Figure(data=go.Heatmap(z=region_category_summary['Sales'],
                                    x=region_category_summary['Region'],
                                    y=region_category_summary['Category']))

fig11.update_layout(title_text='Sales by Region and Category')
fig11.show()

### Scatter Plot

In [65]:
# Using a scatter plot to visualize the relationship between sales and profit
fig11 = px.scatter(df, x='Sales', y='Profit', title='Relationship between Sales and Profit')
fig11.show()

In [66]:
# using a scatter plot to compare the sales and profit by region
fig12 = px.scatter(df, x='Sales', y='Profit', color='Region', title='Sales and Profit by Region')
fig12.show()

### Country Analysis

#### Top Country by Sales and Profit

In [67]:
# Top Customers by Sales
top_customers_sales = df.groupby('Country')['Sales'].sum().reset_index().sort_values(by='Sales', ascending=False).head(10)
fig13 = px.bar(top_customers_sales, x='Sales', y='Country', orientation='h', title='Top 10 Countries by Sales',
              labels={'Sales': 'Total Sales', 'Country': 'Country'}, color='Sales')
fig13.show()

# Top Customers by Profit
top_customers_profit = df.groupby('Country')['Profit'].sum().reset_index().sort_values(by='Profit', ascending=False).head(10)
# Using diffrent nice colors for the bar chart
fig144 = px.bar(top_customers_profit, x='Profit', y='Country', orientation='h', title='Top 10 Countries by Profit',
             labels={'Profit': 'Total Profit', 'Country': 'Country'}, color='Profit', color_continuous_scale='Viridis')
fig14.show()



### Customer Distribution by Segments

In [68]:
# Customer Segmentation by Segment
segment_distribution = df['Segment'].value_counts().reset_index()
segment_distribution.columns = ['Segment', 'count']

fig15 = px.pie(segment_distribution, values='count', names='Segment', title='Customer Segmentation by Segment',
               labels={'Segment': 'Customer Segment', 'count': 'Number of Customers'})
fig15.show()


* Sales by Customer Segment

In [69]:

segment_sales_profit = df.groupby('Segment').agg({'Sales': 'sum', 'Profit': 'sum'}).reset_index()
fig16 = px.bar(segment_sales_profit, x='Segment', y='Sales', title='Sales by Customer Segment',
              labels={'Segment': 'Customer Segment', 'Sales': 'Total Sales'}, color='Segment')
fig16.show()


fig17 = px.bar(segment_sales_profit, x='Segment', y='Profit', title='Profit by Customer Segment',
              labels={'Segment': 'Customer Segment', 'Profit': 'Total Profit'}, color='Segment')
fig17.show()


### Product Analysis

* Product Perfomance

In [70]:
# Top Products by Sales
top_products_sales = df.groupby('Product.Name')['Sales'].sum().reset_index().sort_values(by='Sales', ascending=False).head(5)
fig18 = px.bar(top_products_sales, x='Sales', y='Product.Name', orientation='h', title='Top 5 Products by Sales',
              labels={'Sales': 'Total Sales', 'Product.Name': 'Product Name'}, color='Sales')
fig18.show()

# Top Products by Profit
top_products_profit = df.groupby('Product.Name')['Profit'].sum().reset_index().sort_values(by='Profit', ascending=False).head(5)
fig19 = px.bar(top_products_profit, x='Profit', y='Product.Name', orientation='h', title='Top 5 Products by Profit',
              labels={'Profit': 'Total Profit', 'Product.Name': 'Product Name'}, color='Profit')
fig19.show()


### Time Series

* Seasonal Trends

In [71]:
df['order_date'] = pd.to_datetime(df['Order.Date'])
df['Year_month'] = df['Order.Date'].dt.to_period('M').astype(str)
seasonal_trends = df.groupby('Year_month')['Sales'].sum().reset_index()

fig19 = px.line(seasonal_trends, x='Year_month', y='Sales', title='Seasonal Sales Trends',
               labels={'Year_month': 'Year-Month', 'Sales': 'Total Sales'},
               markers=True)
fig19.update_traces(line_color='blue')
fig19.show()


### Market Analysis

* MArket Comparison

In [72]:
market_comparison = df.groupby('Market')['Sales'].sum().reset_index()
fig20 = px.bar(market_comparison, x='Sales', y='Market', orientation='h', title='Total Sales by Market',
              labels={'Sales': 'Total Sales', 'Market': 'Market'}, color='Sales')
fig20.show()


### Regional Perfomance

In [73]:
regional_preferences = df.groupby(['Region', 'Category'])['Sales'].sum().reset_index()
fig21 = px.bar(regional_preferences, x='Sales', y='Region', color='Category', title='Regional Preferences by Category',
              labels={'Sales': 'Total Sales', 'Region': 'Region'}, barmode='group')
fig21.show()


### Profitabilty Analysis

* Discount Impact

In [74]:
discount_impact = df.groupby('Discount')['Profit'].mean().reset_index()
fig22 = px.line(discount_impact, x='Discount', y='Profit', title='Impact of Discount on Profit',
                labels={'Discount': 'Discount Rate', 'Profit': 'Average Profit'},
                markers=True)
fig22.update_traces(line_color='green')
fig22.show()


### Cost Analysis

* Shipping Perfomance

In [75]:
cost_analysis = df.groupby('Ship.Mode')['Shipping.Cost'].mean().reset_index()
fig23 = px.bar(cost_analysis, x='Shipping.Cost', y='Ship.Mode', orientation='h', title='Average Shipping Cost by Mode',
               labels={'Shipping.Cost': 'Average Shipping Cost', 'Ship.Mode': 'Shipping Mode'}, color='Shipping.Cost')
fig23.show()


### Profit Margin Analysis

In [76]:
# Calculate profit margin
df['profit_margin'] = df['Profit'] / df['Sales']


* Profit Margin by Category

In [77]:
# Profit margin by category
category_profit_margin = df.groupby('Category')['profit_margin'].mean().reset_index()
fig24 = px.bar(category_profit_margin, x='Category', y='profit_margin', title='Profit Margin by Category',
               labels={'Category': 'Category', 'profit_margin': 'Profit Margin'}, color='profit_margin')

fig24.show()

* Profit Margin by Region

In [78]:
# Profit margin by Region
profit_margin_region = df.groupby('Region')['profit_margin'].mean().reset_index()
fig25 = px.bar(profit_margin_region, x='profit_margin', y='Region', orientation='h', title='Profit Margin by Region',
               labels={'profit_margin': 'Profit Margin', 'Region': 'Region'}, color='profit_margin')
fig25.show()

* Profit margin by Region

In [79]:
# Profit Margin Analysis by Region
region_profit_margin = df.groupby('Region').agg({'profit_margin': 'mean'}).reset_index()

# Create visualization
fig26 = px.bar(region_profit_margin, x='profit_margin', y='Region', orientation='h', title='Profit Margin by Region',
              labels={'profit_margin': 'Average Profit Margin', 'Region': 'Region'}, color='profit_margin')
fig26.show()


### Further Insights

* Most and Least Profitable Product

In [80]:
most_profitable_category = category_summary.loc[category_summary['Profit'].idxmax()]
least_profitable_category = category_summary.loc[category_summary['Profit'].idxmin()]

print(f"Most profitable product category: {most_profitable_category['Category']} with average profit of {most_profitable_category['Profit']}")
print(f"Least profitable product category: {least_profitable_category['Category']} with average profit of {least_profitable_category['Profit']}")


Most profitable product category: Technology with average profit of 65.4549584044966
Least profitable product category: Office Supplies with average profit of 16.5789605826112


* Regions with Highest Profits and Sales

In [81]:
highest_sales_region = region_summary.loc[region_summary['Sales'].idxmax()]
highest_profit_region = region_summary.loc[region_summary['Profit'].idxmax()]

print(f"Region with the highest sales: {highest_sales_region['Region']} with total sales of {highest_sales_region['Sales']}")
print(f"Region with the highest average profit: {highest_profit_region['Region']} with average profit of {highest_profit_region['Profit']}")


Region with the highest sales: Central with total sales of 124709
Region with the highest average profit: Central with average profit of 16619.21676


In [82]:
total_sales = df['Sales'].sum()
total_profit = df['Profit'].sum()
total_orders = df['Order.ID'].nunique()

print(f"Total sales: {total_sales}")
print(f"Total profit: {total_profit}")
print(f"Total number of orders: {total_orders}")


Total sales: 12642905
Total profit: 1467457.2912800002
Total number of orders: 25035


# Dash Dashboard

In [83]:
# Insalling dash for creating dashboard

#%pip install dash

In [84]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from dash import Dash, dcc, html, Input, Output

## Dash Dashboard from scratch

In [85]:
# importing the required modules
import pandas as pd  # for data manipulation
import numpy as np  # for numerical computation
import matplotlib.pyplot as plt # for data visualization
import seaborn as sns # for data visualization 
import plotly.express as px # for data visualization
import plotly.graph_objects as go # for data visualization
from plotly.subplots import make_subplots # for data visualization
import os # for file handling
import sys # for system operations
import warnings # for ignoring warnings
warnings.filterwarnings('ignore') # ignoring warnings
from dash import dash_table


In [86]:
from dash import Dash, dcc, html, Input, Output
import plotly.express as px
import plotly.graph_objects as go
import pandas as pd

app = Dash(__name__)

# Load the data
df = pd.read_csv("C:\\Users\\Admin\\OneDrive\\R STA1040\\superstore.csv")
df['Order.Date'] = pd.to_datetime(df['Order.Date'])
df['Year_month'] = df['Order.Date'].dt.to_period('M').astype(str)

#Summarize total sales, average profit, and number of orders by region or category.
region_summary = df.groupby('Region').agg({'Sales': 'sum', 'Profit': 'mean', 'Quantity': 'sum'}).reset_index()



# Visualizing the Region Summary
fig = make_subplots(rows=1, cols=3, subplot_titles=("Total Sales by Region", "Average Profit by Region", "Total Quantity by Region"))

fig.add_trace(go.Bar(x=region_summary['Region'], y=region_summary['Sales'], marker=dict(color=region_summary['Sales'], coloraxis="coloraxis")), row=1, col=1)
fig.add_trace(go.Bar(x=region_summary['Region'], y=region_summary['Profit'], marker=dict(color=region_summary['Profit'], coloraxis="coloraxis")), row=1, col=2)
fig.add_trace(go.Bar(x=region_summary['Region'], y=region_summary['Quantity'], marker=dict(color=region_summary['Quantity'], coloraxis="coloraxis")), row=1, col=3)

fig.update_layout(coloraxis=dict(colorscale='Viridis'), showlegend=False, title_text="Region Summary")

# update the background color of the plot not the bars
fig.update_layout(plot_bgcolor='pink', paper_bgcolor='lightblue', font_color="black")
#fig.show()



# %%
category_summary = df.groupby('Category').agg({'Sales': 'sum', 'Profit': 'mean', 'Quantity': 'count'}).reset_index()
#print("Category Summary:")
category_summary.sort_values(by='Sales', ascending=False)


# %%
# Using a donut chart to visualize the category summary by sales and profit
# indicate which chart is for sales and which chart is for profit by using different colors and labels
fig1 = make_subplots(rows=1, cols=2, specs=[[{'type':'domain'}, {'type':'domain'}]])
fig1.add_trace(go.Pie(labels=category_summary['Category'], values=category_summary['Sales'], name='Sales', marker_colors=px.colors.sequential.RdPu_r), 1, 1)
fig1.add_trace(go.Pie(labels=category_summary['Category'], values=category_summary['Profit'], name='Profit', marker_colors=px.colors.sequential.RdPu_r), 1, 2)
fig1.update_traces(hole=.5, hoverinfo="label+percent+name")
fig1.update_layout(title_text="Category Summary by Sales and Profit")
fig1.update_layout(paper_bgcolor="lightblue")
#fig1.show()


# %%
# Creating a subplot for plotting the sales, profit and quantity by region on pie chart
fig2 = make_subplots(rows=1, cols=3, specs=[[{'type':'domain'}, {'type':'domain'}, {'type':'domain'}]], subplot_titles=('Sales by Region', 'Profit by Region', 'Quantity by Region'))

# Plotting the sales by region

fig2.add_trace(go.Pie(labels=region_summary['Region'], values=region_summary['Sales'], name='Sales'), 1, 1)

# Plotting the profit by region
fig2.add_trace(go.Pie(labels=region_summary['Region'], values=region_summary['Profit'], name='Profit'), 1, 2)

# Plotting the quantity by region
fig2.add_trace(go.Pie(labels=region_summary['Region'], values=region_summary['Quantity'], name='Quantity'), 1, 3)

fig2.update_layout(title_text='Sales, Profit and Quantity by Region')
#fig2.show()


# %% [markdown]
# ### Sales Trend over time

# %%
import plotly.express as px

#df['order_date'] = pd.to_datetime(df['order_date'])
df['Year_month'] = df['Order.Date'].dt.to_period('M').astype(str)
sales_trends = df.groupby('Year_month')['Sales'].sum().reset_index()

fig3 = px.line(sales_trends, x='Year_month', y='Sales', title='Sales Trends Over Time',
               labels={'Year_month': 'Year-Month', 'Sales': 'Total Sales'},
               markers=True)
fig3.update_traces(line_color='blue')
#fig3.show()


# %%
# Create Year_month column
df['Year_month'] = df['Order.Date'].dt.to_period('M').astype(str)

# Region wise sales trends
region_sales_trends = df.groupby(['Year_month', 'Region'])['Sales'].sum().reset_index()

fig4 = px.line(region_sales_trends, x='Year_month', y='Sales', title='Region Wise Sales Trends Over Time',
                color='Region', labels={'Year_month': 'Year-Month', 'Sales': 'Total Sales'},
                markers=True)
#fig4.show()

# %%
# Region wise sales, profit and quantity trends over time in subplots for top 5 regions
top_5_regions = df.groupby('Region')['Sales'].sum().sort_values(ascending=False).head(3).index
region_summary = df.groupby(['Year_month', 'Region']).agg({'Sales': 'sum', 'Profit': 'sum', 'Quantity': 'sum'}).reset_index()

fig5 = make_subplots(rows=1, cols=2, subplot_titles=('Sales Trends', 'Profit Trends'))

# Plotting the sales trends
for region in top_5_regions:
    fig5.add_trace(go.Scatter
                        (x=region_summary[region_summary['Region'] == region]['Year_month'],
                        y=region_summary[region_summary['Region'] == region]['Sales'],
                        mode='lines+markers', name=region), row=1, col=1)
    
# Plotting the profit trends
for region in top_5_regions:
    fig5.add_trace(go.Scatter
                        (x=region_summary[region_summary['Region'] == region]['Year_month'],
                        y=region_summary[region_summary['Region'] == region]['Profit'],
                        mode='lines+markers', name=region), row=1, col=2)
    
# Plotting the quantity trends
#for region in top_5_regions:
    #fig4.add_trace(go.Scatter
                        #(x=region_summary[region_summary['Region'] == region]['Year_month'],
                        #y=region_summary[region_summary['Region'] == region]['Quantity'],
                        #mode='lines+markers', name=region), row=2, col=1)

fig5.update_layout(title_text='Top 3 Region Wise Sales, Profit and Quantity Trends Over Time')
#fig5.show()

# %% [markdown]
# ### Total sales by Category and region

# %%
# Using bar chart to visualize the top 10 products by sales and profit and quantity using a subplot
top_5_products = df.groupby('Product.Name').agg({'Sales': 'sum', 'Profit': 'sum', 'Quantity': 'sum'}).reset_index()
top_5_products = top_5_products.sort_values(by='Sales', ascending=False).head(3)

fig6 = make_subplots(rows=1, cols=3, subplot_titles=('Sales', 'Profit', 'Quantity'))

# Plotting the sales by product
fig6.add_trace(go.Bar(x=top_5_products['Product.Name'], y=top_5_products['Sales'], name='Sales'), 1, 1)

# Plotting the profit by product
fig6.add_trace(go.Bar(x=top_5_products['Product.Name'], y=top_5_products['Profit'], name='Profit'), 1, 2)

# Plotting the quantity by product
fig6.add_trace(go.Bar(x=top_5_products['Product.Name'], y=top_5_products['Quantity'], name='Quantity'), 1, 3)

fig6.update_layout(title_text='Top 5 Products by Sales, Profit and Quantity')
# background color
fig6.update_layout(paper_bgcolor="khaki")
#fig6.show()


# %%
# Total sales, profit and quantity by segment
segment_summary = df.groupby('Segment').agg({'Sales': 'sum', 'Profit': 'sum', 'Quantity': 'sum'}).reset_index()
#segment_summary

# %%
# A plot of total sales and profit by segment as grouped bar chart on a single plot
fig7 = go.Figure(data=[
    go.Bar(name='Sales', x=segment_summary['Segment'], y=segment_summary['Sales']),
    go.Bar(name='Profit', x=segment_summary['Segment'], y=segment_summary['Profit'])
])

# Change the bar mode
fig7.update_layout(barmode='group')
fig7.update_layout(title_text='Total Sales and Profit by Segment')
#fig7.show()

# %%
# Total sales, profit and quantity by Region, Segment and Category after grouping by Region, Segment and Category
region_segment_category_summary = df.groupby(['Region']).agg({'Sales': 'sum', 'Profit': 'sum'}).reset_index()
#region_segment_category_summary


# %%
# A plot of total sales and profit by segment, Region, and Category as grouped bar chart on a single plot
fig8 = go.Figure(data=[
    go.Bar(name='Sales', x=region_segment_category_summary['Region'], y=region_segment_category_summary['Sales']),
    go.Bar(name='Profit', x=region_segment_category_summary['Region'], y=region_segment_category_summary['Profit'])
])

# Change the bar mode
fig8.update_layout(barmode='group')
fig8.update_layout(title_text='Total Sales and Profit by Region')
#fig8.show()

# %%
# Sales by Region and Category using a bar chart and a grouped bar chart and arrange them in a subplot
region_category_summary = df.groupby(['Region', 'Category']).agg({'Sales': 'sum'}).reset_index()

fig9 = make_subplots(rows=1, cols=2, subplot_titles=('Sales by Region and Category', 'Sales by Region and Category'))

# Plotting the sales by region and category
fig9.add_trace(go.Bar(x=region_category_summary['Region'], y=region_category_summary['Sales'], name='Sales'), 1, 1)

# Plotting the sales by region and category as grouped bar chart
fig9.add_trace(go.Bar(x=region_category_summary['Region'], y=region_category_summary['Sales'], name='Sales'), 1, 2)

fig9.update_layout(title_text='Sales by Region and Category')
#fig9.show()


# %% [markdown]
# ### Correlation between Profit and Quantity Sold
# 
# ## HeatMap

# %%
# Using heatmap to visualize the correlation between sales, profit, and quantity, discount, and shipping cost
correlation = df[['Sales', 'Profit','Discount']].corr()
fig10 = go.Figure(data=go.Heatmap(z=correlation.values, x=correlation.columns, y=correlation.columns))
fig10.update_layout(title_text='Correlation between Sales, Profit, Discount')
#fig10.show()

# %%
# Plotting a heatmap to visualize the sales, profit and quantity by Region and Category
region_category_summary = df.groupby(['Region', 'Category']).agg({'Sales': 'sum', 'Profit': 'sum', 'Quantity': 'sum'}).reset_index()

fig11 = go.Figure(data=go.Heatmap(z=region_category_summary['Sales'],
                                    x=region_category_summary['Region'],
                                    y=region_category_summary['Category']))

fig11.update_layout(title_text='Sales by Region and Category')
#fig11.show()

# %% [markdown]
# ### Scatter Plot

# %%
# Using a scatter plot to visualize the relationship between sales and profit
fig11 = px.scatter(df, x='Sales', y='Profit', title='Relationship between Sales and Profit')
#fig11.show()

# %%
# using a scatter plot to compare the sales and profit by region
fig12 = px.scatter(df, x='Sales', y='Profit', color='Region', title='Sales and Profit by Region')
#fig12.show()

# %% [markdown]
# ### Country Analysis
# 
# #### Top Country by Sales and Profit

# %%
# Top Customers by Sales
top_customers_sales = df.groupby('Country')['Sales'].sum().reset_index().sort_values(by='Sales', ascending=False).head(10)
fig13 = px.bar(top_customers_sales, x='Sales', y='Country', orientation='h', title='Top 10 Countries by Sales',
              labels={'Sales': 'Total Sales', 'Country': 'Country'}, color='Sales')
#fig13.show()

# Top Customers by Profit
top_customers_profit = df.groupby('Country')['Profit'].sum().reset_index().sort_values(by='Profit', ascending=False).head(10)
# Using diffrent nice colors for the bar chart
fig14 = px.bar(top_customers_profit, x='Profit', y='Country', orientation='h', title='Top 10 Countries by Profit',
             labels={'Profit': 'Total Profit', 'Country': 'Country'}, color='Profit', color_continuous_scale='Viridis')
#fig14.show()



# %% [markdown]
# ### Customer Distribution by Segments

# %%
# Customer Segmentation by Segment
segment_distribution = df['Segment'].value_counts().reset_index()
segment_distribution.columns = ['Segment', 'count']

fig15 = px.pie(segment_distribution, values='count', names='Segment', title='Customer Segmentation by Segment',
               labels={'Segment': 'Customer Segment', 'count': 'Number of Customers'})
#fig15.show()


# %% [markdown]
# * Sales by Customer Segment

# %%

segment_sales_profit = df.groupby('Segment').agg({'Sales': 'sum', 'Profit': 'sum'}).reset_index()
fig16 = px.bar(segment_sales_profit, x='Segment', y='Sales', title='Sales by Customer Segment',
              labels={'Segment': 'Customer Segment', 'Sales': 'Total Sales'}, color='Segment')
#fig16.show()


fig17 = px.bar(segment_sales_profit, x='Segment', y='Profit', title='Profit by Customer Segment',
              labels={'Segment': 'Customer Segment', 'Profit': 'Total Profit'}, color='Segment')
#fig17.show()


# %% [markdown]
# ### Product Analysis
# 
# * Product Perfomance

# %%
# Top Products by Sales
top_products_sales = df.groupby('Product.Name')['Sales'].sum().reset_index().sort_values(by='Sales', ascending=False).head(5)
fig18 = px.bar(top_products_sales, x='Sales', y='Product.Name', orientation='h', title='Top 5 Products by Sales',
              labels={'Sales': 'Total Sales', 'Product.Name': 'Product Name'}, color='Sales')
#fig18.show()

# Top Products by Profit
top_products_profit = df.groupby('Product.Name')['Profit'].sum().reset_index().sort_values(by='Profit', ascending=False).head(5)
fig18 = px.bar(top_products_profit, x='Profit', y='Product.Name', orientation='h', title='Top 5 Products by Profit',
              labels={'Profit': 'Total Profit', 'Product.Name': 'Product Name'}, color='Profit')
#fig18.show()


# %% [markdown]
# ### Time Series
# 
# * Seasonal Trends

# %%
df['order_date'] = pd.to_datetime(df['Order.Date'])
df['Year_month'] = df['Order.Date'].dt.to_period('M').astype(str)
seasonal_trends = df.groupby('Year_month')['Sales'].sum().reset_index()

fig19 = px.line(seasonal_trends, x='Year_month', y='Sales', title='Seasonal Sales Trends',
               labels={'Year_month': 'Year-Month', 'Sales': 'Total Sales'},
               markers=True)
fig19.update_traces(line_color='blue')
#fig19.show()


# %% [markdown]
# ### Market Analysis
# 
# * MArket Comparison

# %%
market_comparison = df.groupby('Market')['Sales'].sum().reset_index()
fig20 = px.bar(market_comparison, x='Sales', y='Market', orientation='h', title='Total Sales by Market',
              labels={'Sales': 'Total Sales', 'Market': 'Market'}, color='Sales')
#fig20.show()


# %% [markdown]
# ### Regional Perfomance

# %%
regional_preferences = df.groupby(['Region', 'Category'])['Sales'].sum().reset_index()
fig21 = px.bar(regional_preferences, x='Sales', y='Region', color='Category', title='Regional Preferences by Category',
              labels={'Sales': 'Total Sales', 'Region': 'Region'}, barmode='group')
#fig21.show()


# %% [markdown]
# ### Profitabilty Analysis
# 
# * Discount Impact

# %%
discount_impact = df.groupby('Discount')['Profit'].mean().reset_index()
fig22 = px.line(discount_impact, x='Discount', y='Profit', title='Impact of Discount on Profit',
                labels={'Discount': 'Discount Rate', 'Profit': 'Average Profit'},
                markers=True)
fig22.update_traces(line_color='green')
#fig22.show()


# %% [markdown]
# ### Cost Analysis
# 
# * Shipping Perfomance

# %%
cost_analysis = df.groupby('Ship.Mode')['Shipping.Cost'].mean().reset_index()
fig23 = px.bar(cost_analysis, x='Shipping.Cost', y='Ship.Mode', orientation='h', title='Average Shipping Cost by Mode',
               labels={'Shipping.Cost': 'Average Shipping Cost', 'Ship.Mode': 'Shipping Mode'}, color='Shipping.Cost')
#fig23.show()


# %% [markdown]
# ### Profit Margin Analysis

# %%
# Calculate profit margin
df['profit_margin'] = df['Profit'] / df['Sales']


# %% [markdown]
# * Profit Margin by Category

# %%
# Profit margin by category
category_profit_margin = df.groupby('Category')['profit_margin'].mean().reset_index()
fig24 = px.bar(category_profit_margin, x='Category', y='profit_margin', title='Profit Margin by Category',
               labels={'Category': 'Category', 'profit_margin': 'Profit Margin'}, color='profit_margin')

#fig24.show()

# %% [markdown]
# * Profit Margin by Region

# %%
# Profit margin by Region
profit_margin_region = df.groupby('Region')['profit_margin'].mean().reset_index()
fig25 = px.bar(profit_margin_region, x='profit_margin', y='Region', orientation='h', title='Profit Margin by Region',
               labels={'profit_margin': 'Profit Margin', 'Region': 'Region'}, color='profit_margin')
#fig25.show()

# %% [markdown]
# * Profit margin by Region

# %%
# Profit Margin Analysis by Region
region_profit_margin = df.groupby('Region').agg({'profit_margin': 'mean'}).reset_index()

# Create visualization
fig26 = px.bar(region_profit_margin, x='profit_margin', y='Region', orientation='h', title='Profit Margin by Region',
              labels={'profit_margin': 'Average Profit Margin', 'Region': 'Region'}, color='profit_margin')
#fig26.show()


# %% [markdown]
# ### Further Insights

# %% [markdown]
# * Most and Least Profitable Product

# %%
most_profitable_category = category_summary.loc[category_summary['Profit'].idxmax()]
least_profitable_category = category_summary.loc[category_summary['Profit'].idxmin()]

#print(f"Most profitable product category: {most_profitable_category['Category']} with average profit of {most_profitable_category['Profit']}")
#print(f"Least profitable product category: {least_profitable_category['Category']} with average profit of {least_profitable_category['Profit']}")


# %% [markdown]
# * Regions with Highest Profits and Sales

# %%
highest_sales_region = region_summary.loc[region_summary['Sales'].idxmax()]
highest_profit_region = region_summary.loc[region_summary['Profit'].idxmax()]

#print(f"Region with the highest sales: {highest_sales_region['Region']} with total sales of {highest_sales_region['Sales']}")
#print(f"Region with the highest average profit: {highest_profit_region['Region']} with average profit of {highest_profit_region['Profit']}")


# %%
total_sales = df['Sales'].sum()
total_profit = df['Profit'].sum()
total_orders = df['Order.ID'].nunique()

#print(f"Total sales: {total_sales}")
#print(f"Total profit: {total_profit}")
#print(f"Total number of orders: {total_orders}")

## DashBoard

In [87]:
from dash import Dash, html, dcc, dash_table

app = Dash(__name__)

# Sample Markdown text
#markdown_text = '''Analyzing and visualizing sales performance across different regions and product categories.  
#To highlight key sales trends, profitable products, and regional performance.'''

# Tab styling
tabs_styles = {
    'height': '44px',
    'backgroundColor': 'rgb(120, 120, 120)'
}
tab_style = {
    'padding': '10px',
    'fontWeight': 'bold',
    'color': '#4B0082',
    'borderBottom': '2px solid #d6d6d6'
}
tab_selected_style = {
    'borderTop': '3px solid #119DFF',
    'backgroundColor': 'rgb(255, 165, 0)',
    'color': 'black',
    'padding': '10px'
}

# Main layout
app.layout = html.Div(
    style={
        'backgroundColor': '#f0f4f7',
        'fontFamily': 'Arial, sans-serif',
        'margin': '0 auto',
        'maxWidth': '1200px',
        'padding': '20px'
    },
    children=[
        # Navbar
        html.Div(
            style={
                'backgroundColor': '#4B0082',
                'padding': '10px',
                'borderRadius': '5px',
                'marginBottom': '20px',
                'display': 'flex',
                'justifyContent': 'space-between',
                'alignItems': 'center',
            },
            children=[
                html.H1(
                    'Global Superstore Dashboard',
                    style={
                        'color': 'white',
                        'margin': '0',
                        'fontSize': '28px'
                    }
                ),
                html.A(
                    'Learn More',
                    href='#',
                    style={
                        'color': 'white',
                        'fontSize': '16px',
                        'textDecoration': 'none',
                        'padding': '8px 16px',
                        'backgroundColor': '#FF6347',
                        'borderRadius': '5px',
                        'boxShadow': '2px 2px 5px rgba(0, 0, 0, 0.3)'
                    }
                )
            ]
        ),
                # Tabs
            ]
        )



# Main layout
app.layout = html.Div(
    style={
        'backgroundColor': 'rgb(170, 200, 226)',
        'fontFamily': 'Arial, sans-serif',
        'padding': '5px',
        'borderRadius': '10px'
    },
    children=[
        html.Div(
            style={'textAlign': 'center', 'marginBottom': '5px', 'marginTop': '1px'},
            children=[
                html.H1(
                    'Global Superstore Sales Dashboard',
                    style={
                        'color': '#4B0082',
                        'backgroundColor': '#ADD8E6',
                        'padding': '10px',
                        'borderRadius': '5px',
                    }
                ),
                html.P(
                    'Analyzing sales performance across different regions and product categories',
                    style={
                        'fontSize': '20px',
                        'color': '#2F4F4F',
                        'padding': '5px',
                        'marginBottom': '10px'
                    }
                ),
                #dcc.Markdown(
                    #markdown_text,
                    #style={
                        #'fontSize': '16px',
                        #'backgroundColor': '#FFEB3B',
                        #'padding': '10px',
                        #'borderRadius': '5px',
                        #'textAlign': 'center'
                    #}
                
            ],
        ),
        # Tabs
        dcc.Tabs(
            style=tabs_styles,
            children=[
                # Regions Tab
                    dcc.Tab(
                    label='Regions',
                    style=tab_style,
                    selected_style=tab_selected_style,
                    children=[
                        html.Div(
                            style={'display': 'flex', 'flexDirection': 'column', 'marginTop': '20px'},
                            children=[
                    html.Div(
                    style={'display': 'flex', 'justifyContent': 'space-between', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-1',
                            figure=fig13,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        ),
                        dcc.Graph(
                            id='customers-graph-2',
                            figure=fig14,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'display': 'flex', 'justifyContent': 'center', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-3',
                            figure=fig,
                            style={'width': '100%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'marginTop': '20px'},
                    children=[
                        dcc.Markdown(
                            id='graph-explanation',
                            children='''
                            #### Regional Sales Distribution
                            The United States is the key revenue driver, with significantly higher sales than other countries, warranting a continued focus on this market.
                            Australia and France show strong performance as secondary markets, while China and India offer growth opportunities with targeted efforts.
                            

                            The Central region is a critical area, excelling in both total sales and quantity. 
                            However, North Asia, despite lower sales volumes, has the highest profitability, signaling an opportunity to prioritize profitability-driven strategies there.
                            ''',
                            style={'width': '100%', 'border': '1px solid #ddd', 'borderRadius': '5px', 'padding': '10px'}
                        )
                    ]
                )
            ]
        )
    ]
),
                # Cost Analysis Tab
                    dcc.Tab(
                    label='Costs',
                    style=tab_style,
                    selected_style=tab_selected_style,
                    children=[
                        html.Div(
                            style={'display': 'flex', 'flexDirection': 'column', 'marginTop': '20px'},
                            children=[
                    html.Div(
                    style={'display': 'flex', 'justifyContent': 'space-between', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-1',
                            figure=fig22,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        ),
                        dcc.Graph(
                            id='customers-graph-2',
                            figure=fig23,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'display': 'flex', 'justifyContent': 'center', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-3',
                            figure=fig26,
                            style={'width': '100%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'marginTop': '10px'},
                    children=[
                        dcc.Markdown(
                            id='graph-explanation',
                            children='''
                            #### Shipping Costs 
                            Same Day is the most expensive, while Standard Class is the cheapest. Use Standard Class for cost-conscious customers and Same Day for urgent needs.
                            #### Profit Margins by Region
                            Canada and West have the highest profitability—focus on expansion there. Losses in Africa, Southeast Asia and EMEA require cost control or pricing adjustments.
                            #### Impact of Discounts on Profit 
                            Profits decline significantly at discounts above 40%. Limit discounts to 40% or lower and explore alternative promotions to sustain profitability.
                            ''',


                            style={'width': '90%', 'border': '1px solid #ddd', 'borderRadius': '5px', 'padding': '10px'}
                        )
                    ]
                )
            ]
        )
    ]
),
                # Products Tab
                dcc.Tab(
                    label='Products',
                    style=tab_style,
                    selected_style=tab_selected_style,
                    children=[
                        html.Div(
                            style={'display': 'flex', 'flexDirection': 'column', 'marginTop': '20px'},
                            children=[
                    html.Div(
                    style={'display': 'flex', 'justifyContent': 'space-between', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-1',
                            figure=fig18,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        ),
                        dcc.Graph(
                            id='customers-graph-2',
                            figure=fig19,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'display': 'flex', 'justifyContent': 'center', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-3',
                            figure=fig6,
                            style={'width': '100%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'marginTop': '20px'},
                    children=[
                        dcc.Markdown(
                            id='graph-explanation',
                            children='''
                            #### Conclusion
                            Canon Copier leads in profit, while Apple Smart Phone dominates sales and quantity but needs better profit margins. 

                            Seasonal sales peaks highlight opportunities to target high-demand periods, while proactive strategies can address low-sales months.

                            Focusing on profitable products like Cisco and Motorola Smart Phones and optimizing pricing can sustain momentum.
                        
                            ''',
                            style={'width': '95%', 'border': '1px solid #ddd', 'borderRadius': '5px', 'padding': '10px'}
                        )
                    ]
                )
            ]
        )
    ]
),

              # Correlation Tab
            
                dcc.Tab(
                    label='Correlation',
                    style=tab_style,
                    selected_style=tab_selected_style,
                    children=[
                        html.Div(
                            style={'display': 'flex', 'justifyContent': 'space-between', 'marginTop': '20px'},
                            children=[
                                dcc.Graph(
                                    id='products-graph-1',
                                    figure=fig10,
                                    style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                                ),
                                dcc.Graph(
                                    id='products-graph-2',
                                    figure=fig12,
                                    style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                                )
                            ]
                        )
                    ]
                ),

                # Time Series Tab
        
                    dcc.Tab(
                    label='Time Series',
                    style=tab_style,
                    selected_style=tab_selected_style,
                    children=[
                        html.Div(
                            style={'display': 'flex', 'flexDirection': 'column', 'marginTop': '20px'},
                            children=[
                    html.Div(
                    style={'display': 'flex', 'justifyContent': 'space-between', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-1',
                            figure=fig3,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        ),
                        dcc.Graph(
                            id='customers-graph-2',
                            figure=fig4,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'display': 'flex', 'justifyContent': 'center', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-3',
                            figure=fig5,
                            style={'width': '100%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'marginTop': '20px'},
                    children=[
                        dcc.Markdown(
                            id='graph-explanation',
                            children='''
                            #### Time Series Analysis
                            The Central region is the top-performing region both in sales and profits, warranting continued investment.
                            The South region needs investigation to address profit volatility and improve performance, despite steady sales.
                            A deeper look at seasonal trends could reveal opportunities to capitalize on peak periods across all regions
                            ''',
                            style={'width': '95%', 'border': '1px solid #ddd', 'borderRadius': '5px', 'padding': '10px'}
                        )
                    ]
                )
            ]
        )
    ]
),

                # Market Tab
                    dcc.Tab(
                    label='Market',
                    style=tab_style,
                    selected_style=tab_selected_style,
                    children=[
                        html.Div(
                            style={'display': 'flex', 'flexDirection': 'column', 'marginTop': '20px'},
                            children=[
                    html.Div(
                    style={'display': 'flex', 'justifyContent': 'space-between', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-1',
                            figure=fig20,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        ),
                        dcc.Graph(
                            id='customers-graph-2',
                            figure=fig1,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'display': 'flex', 'justifyContent': 'center', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-3',
                            figure=fig21,
                            style={'width': '100%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'marginTop': '20px'},
                    children=[
                        dcc.Markdown(
                            id='graph-explanation',
                            children='''
                            #### Market Analysis
                            APAC and EU should be prioritized for investment, as they drive the majority of total sales.
                            US and LATAM are consistent contributors and might benefit from targeted strategies to boost sales further.
                            EMEA, Africa, and Canada need deeper analysis to explore growth opportunities or reconsider allocation of resources.
                            Furniture and Technology are the top-selling categories in Central, while Office Supplies show potential for growth.
                            More empasis to be given to Central and South regions for Furniture and Technology sales.
                            ''',
                            style={'width': '95%', 'border': '1px solid #ddd', 'borderRadius': '5px', 'padding': '10px'}
                        )
                    ]
                )
            ]
        )
    ]
),


# Customers Tab
dcc.Tab(
    label='Customers',
    style=tab_style,
    selected_style=tab_selected_style,
    children=[
        html.Div(
            style={'display': 'flex', 'flexDirection': 'column', 'marginTop': '20px'},
            children=[
                html.Div(
                    style={'display': 'flex', 'justifyContent': 'space-between', 'marginBottom': '20px'},
                    children=[
                        dcc.Graph(
                            id='customers-graph-1',
                            figure=fig7,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        ),
                        dcc.Graph(
                            id='customers-graph-2',
                            figure=fig15,
                            style={'width': '48%', 'border': '1px solid #ddd', 'borderRadius': '5px'}
                        )
                    ]
                ),
                html.Div(
                    style={'marginTop': '0px'},
                    children=[
                        dcc.Markdown(
                            id='graph-explanation',
                            children='''
                            #### Customer Analysis
                            The consumer customer segment is the most profitable with the highest sales and profit while Home Office segment has the lowest sales and profit.
                            The foucus should be on the consumer segment to increase sales and profit, and the Home Office segment should be analyzed to understand the reasons for low sales and profit.
                            ''',
                            style={'width': '95%', 'border': '1px solid #ddd', 'borderRadius': '5px', 'padding': '0px'}
                        )
                    ]
                )
            ]
        )
    ]
),
# Descriptives Tab
                dcc.Tab(
                    label='Descriptives',
                    style=tab_style,
                    selected_style=tab_selected_style,
                    children=[
                        html.Div(
                            style={'display': 'flex', 'justifyContent': 'space-between', 'marginTop': '20px'},
                            children=[
                                html.Div(
                                    style={'width': '48%', 'display': 'flex', 'flexDirection': 'column', 'gap': '20px'},
                                    children=[
                                        dash_table.DataTable(
                                            id='segment-summary-table',
                                            columns=[{"name": i, "id": i} for i in segment_summary.columns],
                                            data=segment_summary.to_dict('records'),
                                            style_table={'height': '126px', 'overflowY': 'auto', 'border': '1px solid #ddd'},
                                            style_cell={'textAlign': 'center'},
                                            style_header={
                                                'backgroundColor': '#4B0082',
                                                'color': 'white',
                                                'fontWeight': 'bold',
                                                'textAlign': 'center'
                                            }
                                        ),
                                        dash_table.DataTable(
                                            id='category-summary-table',
                                            columns=[{"name": i, "id": i} for i in category_summary.columns],
                                            data=category_summary.to_dict('records'),
                                            style_table={'height': '126px', 'overflowY': 'auto', 'border': '1px solid #ddd'},
                                            style_cell={'textAlign': 'center'},
                                            style_header={
                                                'backgroundColor': '#4B0082',
                                                'color': 'white',
                                                'fontWeight': 'bold',
                                                'textAlign': 'center'
                                            }
                                        )
                                    ]
                                ),
                                html.Div(
                                    style={'width': '48%'},
                                    children=[
                                        dash_table.DataTable(
                                            id='region-summary-table',
                                            columns=[{"name": i, "id": i} for i in region_summary.columns],
                                            data=region_summary.to_dict('records'),
                                            style_table={'height': '422px', 'overflowY': 'auto', 'border': '1px solid #ddd'},
                                            style_cell={'textAlign': 'center'},
                                            style_header={
                                                'backgroundColor': '#4B0082',
                                                'color': 'white',
                                                'fontWeight': 'bold',
                                                'textAlign': 'center'
                                            }
                                        )
                                    ]
                                )

                            ]
                        )
                    ]
            ),
                dcc.Tab(
                    label='Insights',
                    style=tab_style,
                    selected_style=tab_selected_style,
                    children=[
                         html.Div(
                    style={'marginTop': '0px'},
                    children=[
                        dcc.Markdown(
                            id='graph-explanation',
                            children='''
                            ## Key Insights:
                            **Regional Focus:** Focus on the United States for revenue generation but explore growth in emerging markets like India and China.

                            **Cost Efficiency:** Leverage cost-effective shipping methods and limit discounts to 40% to protect profitability.

                            **Product Optimization:** Prioritize high-profit products like Canon Copier while enhancing sales strategies for high-volume products like Apple Smart Phones.

                            Customer Segmentation: Focus on the consumer segment, which is the most profitable, while investigating the underperformance of the Home Office segment.

                           ## Recommendations:
                            **Expand Regional Focus:** Invest further in APAC and EU regions while exploring growth strategies for Africa and EMEA.

                            **Optimize Product Sales:** Develop targeted campaigns for high-performing products and optimize pricing strategies for those with strong sales potential.

                            **Cost Management:** Focus on controlling shipping costs and reducing excessive discounting to maintain profitability.
                            ''',
                            style={'width': '95%', 'border': '1px solid #ddd', 'borderRadius': '10px', 'padding': '5px'}
                        )
                    ]
                )
                    ]
                )
            ]
        
        )
    ]
)

if __name__ == '__main__':
    app.run_server(debug=True, port=8053)


---------------------------------------------------------------------------
DuplicateIdError                          Traceback (most recent call last)
DuplicateIdError: Duplicate component id found in the initial layout: `customers-graph-1`

